### PR TITLE
Support ssh-agent and external sign binaries for commit signing with ssh

### LINKED ISSUE #2188: Cant use commit signing
**Describe the bug**
When performing a commit I get the error 
`sign builder error: Failed to retrieve 'user.signingkey' from the git configuration: Currently, we only support a pair of ssh key in disk.`

This is likely because of my git configuration for commit signing.

**To Reproduce**
1. Add the following to your `~/.gitconfig`

```
[user]
...

[gpg]
	format = ssh
[gpg "ssh"]
	program = /Applications/1Password.app/Contents/MacOS/op-ssh-sign
[commit]
	gpgsign = true
[push]
	autoSetupRemote = true
```
2. try to perform a commit
3. See the error above

**Expected behavior**
This should work without error. `git` on the CLI works fine without error

**Screenshots**
Screenshot 2024-04-15 at 11 38 03 AM

**Context (please complete the following information):**
 - M2 MacOS 14.4.1
 - GitUI Version gitui nightly 2024-04-14
 - Rust version: cargo 1.77.1 (e52e36006 2024-03-26)

**Additional context**

### PR DESCRIPTION
**Pull request by chirpcel**
This Pull Request fixes/closes #2188.

It changes the following:
- Respect configuration for alternative signing binary (for example 1Password ssh agent)
- Using ssh-keygen for ssh signing operation to support ssh-agent
  - additional: sharing same implementation for default & configured binary

I followed the checklist:
- [x] I added unittests
- [x] I ran `make check` without errors
- [x] I tested the overall application
- [x] I added an appropriate item to the changelog

### CODE REVIEW

**Comment by chirpcel** [comment]
There's one drawback. Like before, we don't support encrypted private keys which are not loaded to agent. The error message isn't as clear as before, but from a functionality point of view it isn't different. I would make a new issue for this behavior and take a look into supporting encrypted private keys.

**Comment by extrawurst** [comment]
> The error message isn't as clear as before

**Comment by extrawurst** [comment]
> The error message isn't as clear as before

cant we make it as clear as before and then followup 

> supporting encrypted private keys.

**Comment by chirpcel** [comment]
Actually I didn't figured out how to capture the passphrase prompt, as it's spawned directly on the tty.
But I just have an Idea, maybe we could provide an empty passphrase as arg to skip the password prompt and map the "wrong passphrase" error to the previous error message.
I'll take a look at it today.

**Comment by chirpcel** [comment]
@extrawurst Worked, we now have a meaningful error message. I tested with 1Password SSH-Agent, Default SSH-Agent, non encrypted keys on disk and encrypted keys on disk.

**Comment by DaRacci** [comment]
Any movement on this?

This would allow me to use gitui without needing to fallback to shell to actually commit.

**Comment by chirpcel** [comment]
@DaRacci Think maintainer is currently busy as asking for co-maintainers and migrating his repo to an org for better support by more people.
For now you can compile yourself with applied pr as I do or wait.

File: asyncgit/src/sync/sign.rs [hunk 1]
```diff
@@ -261,79 +267,99 @@ impl Sign for GPGSign {
...
+		let mut child = cmd
+			.spawn()
+			.map_err(|e| SignError::Spawn(e.to_string()))?;
+
+		let mut stdin = child.stdin.take().ok_or(SignError::Stdin)?;
+
+		stdin
+			.write_all(commit)
+			.map_err(|e| SignError::WriteBuffer(e.to_string()))?;
+		drop(stdin);
```
Comment (naseschwarz) [inline asyncgit/src/sync/sign.rs:320]: This is fine and simple for now, let's keep it like this. I just want to voice one thought so it can be referenced later: This will need to be changed if we ever support password prompts without a custom askpass. The canonical way for this is to write the commit to a temporary file, sign that with ssh-keygen and read the signature from `tempfilename` + `".sig"`.

File: asyncgit/src/sync/sign.rs [hunk 2]
```diff
@@ -261,79 +267,99 @@ impl Sign for GPGSign {
...
 	#[cfg(test)]
-	fn signing_key(&self) -> &String {
-		&self.signing_key
+	fn signing_key(&self) -> String {
+		self.signing_key.clone()
 	}
 }
 
-/// Sign commit data using `SSHDiskKeySign`
+/// Sign commit data using `SSHSign`
```
Comment (naseschwarz) [inline asyncgit/src/sync/sign.rs:280]:
```suggestion
/// Sign commit data using `ssh-keygen`
```

File: asyncgit/src/sync/sign.rs [hunk 3]
```diff
@@ -261,79 +267,99 @@ impl Sign for GPGSign {
...
+			.stdout(Stdio::piped())
+			.stderr(Stdio::piped())
+			.arg("-Y")
+			.arg("sign")
+			.arg("-n")
+			.arg("git")
+			.arg("-f")
+			.arg(&self.signing_key);
+
+		if &self.program == "ssh-keygen" {
```
Comment (naseschwarz) [inline asyncgit/src/sync/sign.rs:305]: Sorry, I don't really understand why you only pass `-P` if this condition is fulfilled.

**Review by naseschwarz — CHANGES_REQUESTED** [review]
Hey @chirpcel. Thank you for this implementation. I support this fully and agree that this should be the way commit signing is implemented at the moment. It is almost in compliance with `git commit`'s implementation. Where changes will be required for currently unsupported `git commit` signing features, I have remarked them in this review. I wouldn't say that these are blockers, though.

I also have minor questions and nitpicks, mostly about corner cases, here and there and some FYIs.

On which platforms have you tested this? I have tested it on Linux and it works as promised. It would be great if someone tested this on Windows and Mac, too, though. (A test should consist of at least signing a commit and checking whether that commit's signature can be validated.)

File: asyncgit/src/sync/sign.rs (hunk 3 above)
Comment (chirpcel) [reply asyncgit/src/sync/sign.rs:305]: Cause there are some implementations which doesn't support empty -P with fallback to agent. For example the 1Password SSH-Agent and it's cli `op-sign` throws an error in that case. On default `ssh-keygen` its tested to work but it hasn't to be on 3rd parties.

File: asyncgit/src/sync/sign.rs [hunk 4]
```diff
@@ -261,79 +267,99 @@ impl Sign for GPGSign {
...
+			cmd.arg("-P").arg("\"\"");
+		}
```
[the remark this answers is not in the document]
Comment (chirpcel) [reply asyncgit/src/sync/sign.rs:307]: Seems like a more clean solution for me. I'll like to change it to that way

**Comment by chirpcel** [comment]
Thanks @naseschwarz for your cr! I commited your suggestions and commented on questions. I'll take a look at the remaining annotations of edge cases/optimizations. I'll update the pr soon. 

I've tested the implementation on linux and macOS. Windows currently missing, maybe I can test it in a VM. Currently don't know about license/activation. Keep you updated.

**Comment by hazzuk** [comment]
Would simply like to add that I've been been running a compiled version of gitui with this pull request since August. So far I've had no issues on my Linux setup.

For me, ssh-agent support was critical, so I'd also like to say thank you for this work. I hope we see this merged one day.

**Comment by chirpcel** [comment]
@hazzuk Me too and I totally forgot that I have this PR open :face_in_clouds:  
You just bring back my attention to it. I'll sort out the last open notes and the we should get it merged.

File: asyncgit/src/sync/sign.rs [hunk 5]
```diff
@@ -156,35 +155,52 @@ impl SignBuilder {
...
+			let mut temp_file =
+				NamedTempFile::new().map_err(|err| {
+					SignBuilderError::SSHSigningKey(err.to_string())
+				})?;
+			writeln!(temp_file, "{signing_key}").map_err(|err| {
+				SignBuilderError::SSHSigningKey(err.to_string())
+			})?;
+			let temp_file = temp_file.keep().map_err(|err| {
+				SignBuilderError::SSHSigningKey(err.to_string())
+			})?;
```
[the remark this answers is not in the document]
Comment (chirpcel) [reply asyncgit/src/sync/sign.rs]: I hold the instance now until drop of SSHSign struct

File: asyncgit/src/sync/sign.rs [hunk 6]
```diff
@@ -156,35 +155,52 @@ impl SignBuilder {
...
 			}
 			_ => Err(SignBuilderError::InvalidFormat(format)),
 		}
 	}
+
+	fn signing_key_into_path(
+		signing_key: &str,
+	) -> Result<PathBuf, SignBuilderError> {
+		let key_path = PathBuf::from(signing_key);
+		if signing_key.starts_with("ssh-") {
```
[the remark this answers is not in the document]
Comment (chirpcel) [reply asyncgit/src/sync/sign.rs]: I've handled this case now. If there's a `key::` prefix I remove it for further processing.

File: asyncgit/src/sync/sign.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (chirpcel) [reply asyncgit/src/sync/sign.rs:360]: I switched to the documented `-U` flag.

**Comment by chirpcel** [comment]
Open notes should now been solved. 
I'll test tomorrow on some machines with different configs and then request an follow-up review.

**Comment by chirpcel** [comment]
I didn't encounter problems on my machines. Tested on linux with openssh agent and pgp based ssh agent.

**Comment by asgeirn** [comment]
Tested on MacOS using 1Password signing, worked perfectly!

**Comment by chirpcel** [comment]
@extrawurst 
As this initial PR is some months old and it happened much since them, who is currently able to do a review on changes?

**Comment by cruessler** [comment]
I’ll have a look!

**Comment by cruessler** [comment]
(Just wanted to let you know that I haven’t forgotten this PR. I hope to be getting to it over the weekend.)

File: asyncgit/src/sync/sign.rs [hunk 7]
```diff
@@ -156,35 +155,62 @@ impl SignBuilder {
...
-						SignBuilderError::SSHSigningKey(String::from(
-							"ssh key setting absent",
-						))
-					})
-					.and_then(SSHSign::new)?;
-				let signer: Box<dyn Sign> = Box::new(ssh_signer);
-				Ok(signer)
+					.map(|value| {
+						if value.starts_with("key::") {
+							value.replacen("key::", "", 1)
```
Comment (cruessler) [inline asyncgit/src/sync/sign.rs]: What do you think about `value.strip_prefix("key::")` instead (seems more idiomatic)?

File: asyncgit/src/sync/sign.rs [hunk 8]
```diff
@@ -156,35 +155,62 @@ impl SignBuilder {
...
+					let named_temp_file =
+						Self::signing_key_into_temp_file(
+							signing_key.as_str(),
+						)?;
+					signing_key = format!(
+						"{}",
+						named_temp_file.path().display()
+					);
+					pub_key_temp_file = Some(named_temp_file);
+				}
```
Comment (cruessler) [inline asyncgit/src/sync/sign.rs]: What do you think about changing to something like the following:

```
let (signing_key, pub_key_temp_file) = {
  …
};
```

This could make the data flow easier to follow.

File: asyncgit/src/sync/sign.rs [hunk 9]
```diff
@@ -156,35 +155,62 @@ impl SignBuilder {
 				String::from("x509"),
 			)),
 			"ssh" => {
-				let ssh_signer = config
+				let program = config
+					.get_string("gpg.ssh.program")
+					.unwrap_or_else(|_| "ssh-keygen".to_string());
+
+				let mut signing_key = config
```
Comment (cruessler) [inline asyncgit/src/sync/sign.rs]: What do you think about renaming to `signing_key_path` since that’s the field name in `SSHSign`?

File: asyncgit/src/sync/sign.rs [hunk 10]
```diff
@@ -156,35 +155,62 @@ impl SignBuilder {
...
+					program,
+					signing_key_path: signing_key,
+					_pub_key_temp_file: pub_key_temp_file,
+				}))
 			}
 			_ => Err(SignBuilderError::InvalidFormat(format)),
 		}
 	}
+
+	fn signing_key_into_temp_file(
```
Comment (cruessler) [inline asyncgit/src/sync/sign.rs]: What do you think about prefixing this with a verb so that it becomes `write_signing_key_to_temp_file` (or something similar)?

File: asyncgit/src/sync/sign.rs [hunk 11]
```diff
@@ -261,79 +277,90 @@ impl Sign for GPGSign {
...
-					))
-				})
-		} else {
-			Err(SignBuilderError::SSHSigningKey(
-				String::from("Currently, we only support a pair of ssh key in disk."),
-			))
-		}
-	}
+	signing_key_path: String,
+	_pub_key_temp_file: Option<NamedTempFile>,
```
Comment (cruessler) [inline asyncgit/src/sync/sign.rs:333]: Why is this prefixed with `_`? Does this warrant a comment?

**Review by cruessler — CHANGES_REQUESTED** [review]
First of all, sorry for the long delay! (I wanted to make sure I had the time and focus needed for an actual review, and that took some time, unfortunately.)

Overall, this looks good to me, as far as I can tell. I just have a few questions.

**Comment by SergioRibera** [comment]
Any movement on this? 👀

**Comment by asgeirn** [comment]
With some help from Claude I've rebased this branch on the latest master, it's at https://github.com/asgeirn/gitui/tree/feature/external-ssh-signer for those who want to try it.

**Comment by extrawurst** [comment]
@asgeirn are you interested in driving this PR over the finish line? i adopted it. could use some folks actually using ssh-agent based signing to give it another go and test

**Comment by extrawurst** [comment]
@cruessler I adopted this PR and cleaned up the remaining nits. can you give it a last test run 🙏

**Comment by asgeirn** [comment]
Tested it with my setup (1password agent commit signing) and it works.
> warning: gitui@0.28.1: buildname '0.28.1-nightly 2026-07-01 (ca34d08)'

**Comment by cruessler** [comment]
> @cruessler I adopted this PR and cleaned up the remaining nits. can you give it a last test run 🙏

Will do!

**Review by cruessler — CHANGES_REQUESTED** [review]
I just tested this branch on my Linux machine (Ubuntu 26.04), configuring a new repo as follows:

```bash
git init
echo "test" > test.txt
git config set --local gpg.format ssh
git config set --local user.signingKey "..."
git config set --local commit.gpgsign true
```

Then, using the binary built via `cargo build`, I tried to commit, but the command seemed to hang indefinitely. This was the last line printed to the log:

```
18:36:13 [TRACE] (1) asyncgit::sync::sign: [asyncgit/src/sync/sign.rs:362] signing command: "ssh-keygen" "-Y" "sign" "-n" "git" "-f" "/tmp/.tmpskBtps" "-U"
```

Running the `ssh-keygen` command outside of `gitui` resulted in:

```
ssh-keygen -Y sign -n git -f /tmp/.tmpskBtps -U
Signing data on standard input
```

Also, I got two `ssh-add` processes running at about 100 % CPU in the background.

When I ran `ssh-add` before committing, though, `gitui` did not hang and `git cat-file commit HEAD` showed that the commit had been signed as it had `BEGIN SSH SIGNATURE` in the `gpgsig` header.

**Comment by extrawurst** [comment]
@cruessler i created a reproducer test for your finding. please verify it works in the above case for you now. did you have any other findings or concerns that we should hold out a merge for?

**Comment by cruessler** [comment]
I just checked again and it worked without me having to run `ssh-add` before. It’s possible that I used the wrong format for `user.signingKey` initially, though I don’t know for sure. No objections from my side!

**Review by cruessler — APPROVED** [review]

**Comment by extrawurst** [comment]
@chirpcel congrats for landing this bad boy! Thanks everyone for your reviews and testing @naseschwarz, @cruessler, @asgeirn - lets test the shit out of the next nightly!
